### ISSUE TITLE
ContentType.JSON should automatically encode as UTF-8

### ISSUE THREAD

**Issue by johanhaleby** [jayway/rest-assured#567]
_From [mchen...@gmail.com](https://code.google.com/u/102335645129723678872/) on June 25, 2015 06:42:43_

What steps will reproduce the problem? 1. Set the contentType of a request to `ContentType.JSON`
    1.b do NOT set the character encoding
2. Add a Unicode body with characters not within ISO-8859-1
3. Send the data. What is the expected output? What do you see instead? - Expect: the data to be sent as specified.

 - Actual: non ISO-8859-1 characters are mangled. What version of the product are you using? On what operating system? 2.4.1, OSX 10.9, 10.10 Please provide any additional information below. When sending JSON data, any non ISO-8859-1 characters are mangled unless the encoding is specified as UTF-8. Issue `#228` initially brought up the issue, but the solution to specify a charset of the content-type, while useful for others, is unnecessary for JSON, and arguably wrong.

From rfc4627 : http://www.ietf.org/rfc/rfc4627.txt JSON text SHALL be encoded in Unicode.  The default encoding is UTF-8.

   Since the first two characters of a JSON text will always be ASCII
   characters [ RFC0020 ], it is possible to determine whether an octet
   stream is UTF-8, UTF-16 (BE or LE), or UTF-32 (BE or LE) by looking
   at the pattern of nulls in the first four octets.

Accordingly, charset parameter is not allowed in the JSON MIME type specifically because the unicode variant can be determined from the content:

The MIME media type for JSON text is application/json.

   Type name: application

   Subtype name: json

   Required parameters: n/a

   Optional parameters: n/a

   Encoding considerations: 8bit if UTF-8; binary if UTF-16 or UTF-32

      JSON may be represented using UTF-8, UTF-16, or UTF-32.  When JSON
      is written in UTF-8, JSON is 8bit compatible.  When JSON is
      written in UTF-16 or UTF-32, the binary content-transfer-encoding
      must be used.

The preferred workaround should be to set the defaultContentCharset on the encoderConfig(), and avoid unnecessary charset parameter.

    newConfig().encoderConfig(encoderConfig().defaultContentCharset("UTF-8"))

_Original issue: http://code.google.com/p/rest-assured/issues/detail?id=412_

**Comment by johanhaleby** [comment]
UTF-8 was the default charset in previous version of REST Assured but this was changed to the default charset of the host. All content is encoded this way. Perhaps we need a way to set not only "default content charset" but also a default content charset for a specific content-type and fallback to "default content charset" if there's no match.

**Comment by johanhaleby** [comment]
I've added support for specifying a default charset for individual content-types now using the `EncoderConfig` (`defaultCharsetForContentType(..)`). By default "application/json" uses UTF-8.

Please try it out be depending on version 2.5.1-SNAPSHOT after having added the following Maven repo:

```xml
<repositories>
        <repository>
            <id>sonatype</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots/</url>
            <snapshots />
        </repository>
</repositories>
```

I'll close this issue for now. Let me know if I should reopen it.

**Comment by tazle** [comment]
Which releases should this be fixed in?

**Comment by johanhaleby** [comment]
Looks like 2.6.0 (see [change log](https://raw.githubusercontent.com/rest-assured/rest-assured/master/changelog.txt))

**Comment by garretwilson** [comment]
> Perhaps we need a way to set not only "default content charset" but also a default content charset for a specific content-type and fallback to "default content charset" if there's no match.

There should be no "default charset". Each request should specify its charset. This is a huge oversight that many developers make, because they just merrily assume that whatever default charset is in place will work, when usually it's the system charset, which varies from system to system. Then odd bugs crop up months down the road because if the system works it works just out of luck.

See the bug I filed #926 .

**Comment by tazle** [comment]
For JSON the case for UTF-8 as the default character set is pretty clear. See https://tools.ietf.org/html/rfc7159 section 8.1.

**Comment by garretwilson** [comment]
I don't think that JSON having a "default" charset means that HTTP should not indicate one. Probably your point is that REST Assured provides UTF-8 by default as a convenience, so that developers don't have to explicitly indicate it via the API. I can't say that's incorrect, but I fear it lulls developers into a sense that they never have to indicate the charset.

This is already a big problem. I cannot tell you how many times my students have opened input streams and blindly wrapped them in an `InputStreamReader` without caring what the charset was; or in a unit test extracted bytes from a string without specifying the encoding. It seemed to work on their system, but would break on another system.

Beyond JSON, I think we already have a problem here in REST Assured. Rather than going on here in this closed issue, I'll comment further in #926.
